### PR TITLE
Initial build review

### PR DESCRIPTION
**Pull request by andrewrubin**
Just wanted to get everyone's eyes on this. Disregard the `changed files` on this branch—just have a look over the whole thing and let me know your thoughts. The readme will be the main reference for usage. A few examples in there as well. Thank you!!

### CODE REVIEW

**Comment by liamegan** [comment]
Because we have nothing to review against (THANKS Andrew) I'll just make a bunch of comments.

**Comment by liamegan** [comment]
This is great, really like the opinions expressed, the class layout and the documentation. By god, the documentation makes me happy.

**Comment by liamegan** [comment]
I know the options are fairly sparse at the moment, but I'd really like to move to a destructuring pattern for options for all of our classes. In this cas this would look something like:

```
constructor(element, {
  baseClass = "horizontal-scroll"
} = {}) {
```

The advantage of this pattern is that we can both assign defaults and not worry about order. In this case we would then probably update the body of the constructor to look something like this:

```
  constructor(element, {
      baseClass = "horizontal-scroll"
    } = {}) {
    this.element = element;

    // Grab the data-attributes
    const {
      baseClassName,
      navigation,
      navigationLabel,
      navigationHiddenTextPrev,
      navigationHiddenTextNext,
      scrollIncrement,
      scrollSnap,
    } = this.element.dataset;

    // Set options based on the data-attributes (where applicable).
    // Anything passed to the constructor will take priority.
    this.options = {
      baseClassName: baseClassName || baseClass || "horizontal-scroll",
      easingFunction: (x) =>
        x < 0.5 ? 2 * x * x : 1 - Math.pow(-2 * x + 2, 2) / 2,
      navigation: navigation && navigation === "false" ? false : true,
      navigationLabel: navigationLabel || "Horizontal scroll",
      navigationHiddenTextNext: navigationHiddenTextNext || "Scroll forwards",
      navigationHiddenTextPrev: navigationHiddenTextPrev || "Scroll backwards",
      navigationVisualContentNext: this.element.querySelector(
        ".horizontal-scroll__nav-markup-next"
      )
        ? this.element.querySelector(".horizontal-scroll__nav-markup-next")
            .innerHTML
        : "<span aria-hidden-'true'>></span>",
      navigationVisualContentPrev: this.element.querySelector(
        ".horizontal-scroll__nav-markup-previous"
      )
        ? this.element.querySelector(".horizontal-scroll__nav-markup-previous")
            .innerHTML
        : "<span aria-hidden-'true'><</span>",
      scrollIncrement: (scrollIncrement && parseInt(scrollIncrement)) || 1,
      scrollSnap: scrollSnap === "true" || false
    };
   ... 
```

In this way we're not polluting this options object with anything that might cause problems later on.

**Comment by liamegan** [comment]
I would love to see a little more inline documentation in this class constructor. Nothing major, just what each step is accomplishing. It's currently a bit sparse.

**Comment by liamegan** [comment]
That's all I got! Good stuff, Andrew. Thanks!

**Comment by andrewrubin** [comment]
> I'd really like to move to a destructuring pattern for options for all of our classes

I totally agree, especially about not polluting the `options` object with who-knows-what. I am concerned that this will become difficult to understand really quickly though, since we'd need different variable names for the `options` properties than for the `data-` attribute properties. As you showed in the example, we can't destructure both of these and use the same variable names. The only solution I can think of is to alias/reassign their names when destructuring.

The options can be pretty verbose actually (looks like I didn't document them in the JS file yet 😱 ) so I want to make sure we settle on a solution here before I make any moves.

In my opinion, the benefit of having an unpolluted object is not worth this confusion. Change my mind though! Maybe I'm missing something obvious @liamegan

**Comment by andrewrubin** [comment]
> I would love to see a little more inline documentation in this class constructor. Nothing major, just what each step is accomplishing. It's currently a bit sparse.

Done!

**Comment by liamegan** [comment]
> > I'd really like to move to a destructuring pattern for options for all of our classes
> 
> I totally agree, especially about not polluting the `options` object with who-knows-what. I am concerned that this will become difficult to understand really quickly though, since we'd need different variable names for the `options` properties than for the `data-` attribute properties. As you showed in the example, we can't destructure both of these and use the same variable names. The only solution I can think of is to alias/reassign their names when destructuring.
> 
> The options can be pretty verbose actually (looks like I didn't document them in the JS file yet 😱 ) so I want to make sure we settle on a solution here before I make any moves.
> 
> In my opinion, the benefit of having an unpolluted object is not worth this confusion. Change my mind though! Maybe I'm missing something obvious @liamegan

I'm not quite sure what you mean here. Let's maybe talk through this so you can show me - currently I'm seeing the following options in the source:

- scrollSnap 
- navigation
- navigationLabel
- baseClassName
- scrollIncrement
- easingFunction

**Comment by andrewrubin** [comment]
Alrighty @liamegan that's all refactored to behave as we discussed. Updated the Codepen demos from the readme to reflect these changes.

File: index.js [hunk 1]
```diff
@@ -7,95 +7,144 @@ import tween from "wtc-tween";
...
+      scrollSnap = element.dataset.scrollSnap === "true" || false,
+    } = {}
+  ) {
+    // Assign a reference to our instance's DOM element:
     this.element = element;
 
-    // Grab the data-attributes
-    const {
+    // Assign a reference to our instance's options:
+    this.options = {
```
Comment (liamegan) [inline index.js]: Small thing, but at this point, the options object is redundant. These can all just become class fields, which will also provide more future flexibility.

File: index.js [hunk 2]
```diff
@@ -7,95 +7,144 @@ import tween from "wtc-tween";
...
-  constructor(element, options = {}) {
+  constructor(
+    element,
+    {
+      baseClassName = element.dataset.baseClassName || "horizontal-scroll",
+      // Ease curve defaults to a quadratic in-out:
+      easingFunction = (x) =>
+        x < 0.5 ? 2 * x * x : 1 - Math.pow(-2 * x + 2, 2) / 2,
+      // Navigation defaults to true, unless "false" data-attr is passed:
+      navigation = element.dataset.navigation &&
```
Comment (liamegan) [inline index.js]: This can just become:
```
navigation = element.dataset.navigation !== "false"
```

File: index.js [hunk 3]
```diff
@@ -7,95 +7,144 @@ import tween from "wtc-tween";
...
+   * @param {number} [options.scrollIncrement] - The amount of child items to
+   * advance when using the next/previous navigation buttons.
+   * @param {boolean} [options.scrollSnap] - Whether or not to use CSS
+   * scroll-snapping to auto-align each item when the user stops scrolling.
+   * Cannot be used with `navigation: true`, due to conflicting animations.
    */
-  constructor(element, options = {}) {
+  constructor(
+    element,
+    {
```
Comment (liamegan) [inline index.js:55]: This looks great, thanks!

**Review by liamegan — CHANGES_REQUESTED** [review]
A couple of small changes. Otherwise looks great!

File: index.js (hunk 1 above)
Comment (andrewrubin) [reply index.js]: Cool, added this. Have a look and let me know if I've done what you're expecting. Essentially I just declared all the fields (as undefined), then assigned them inside the constructor.

File: index.js (hunk 2 above)
Comment (andrewrubin) [reply index.js]: Donezo

File: index.js (hunk 1 above)
Comment (andrewrubin) [reply index.js]: Just an update on this @liamegan: I refactored it a bit to set some defaults on the class fields up-front, then reassign them **only if** they had corresponding properties passed in. A lot of if statements to check for those, but I wasn't sure of another solution. Fully tested each one though, and it's all working great. Lemme know your thoughts, thanks!

Comment (liamegan) [reply index.js]: Looks great, thanks!

**Review by liamegan — APPROVED** [review]
Looks good, please merge whenever you're ready.

**Comment by andrewrubin** [comment]
Great, thanks @liamegan!

**Pull request merged**